apps/aefate/src/aefa_engine_state.erl

```diff
@@ -863,3 +924,93 @@ vm_version(#es{vm_version = X}) ->
+breakpoints(#es{debug_info = #debug_info{breakpoints = Breakpoints}}) ->
+    Breakpoints.
+
+%%%------------------
+
+-spec current_instruction(state()) -> integer().
+current_instruction(#es{debug_info = #debug_info{current_instruction = Current}}) ->
+    Current.
+
+-spec inc_current_instruction(state()) -> state().
+inc_current_instruction(ES = #es{debug_info = DbgInfo}) ->
+    ES#es{debug_info = DbgInfo#debug_info{current_instruction = DbgInfo#debug_info.current_instruction + 1}}.
+
+-spec reset_current_instruction(state()) -> state().
+reset_current_instruction(ES = #es{debug_info = DbgInfo}) ->
+    ES#es{debug_info = DbgInfo#debug_info{current_instruction = 0}}.
+
+%%%------------------
+
+-spec debugger_status(state()) -> debugger_status() | disabled.
+debugger_status(#es{debug_info = disabled}) ->
+    disabled;
+debugger_status(#es{debug_info = #debug_info{status = Status}}) ->
+    Status.
+
+-spec set_debugger_status(debugger_status(), state()) -> state().
+set_debugger_status(Status, ES = #es{debug_info = DbgInfo}) ->
+    ES#es{debug_info = DbgInfo#debug_info{status = Status}}.
+
+%%%------------------
+
+-spec debugger_location(state()) -> debugger_location().
+debugger_location(#es{debug_info = #debug_info{location = Location}}) ->
+    Location.
+
+-spec set_debugger_location(debugger_location(), state()) -> state().
+set_debugger_location(Location, ES = #es{debug_info = DbgInfo}) ->
+    ES#es{debug_info = DbgInfo#debug_info{location = Location}}.
+
+%%%------------------
+
+-spec dbg_call_stack(state()) -> [{string(), pos_integer()}].
+dbg_call_stack(#es{debug_info = #debug_info{call_stack = Stack}}) ->
+    Stack.
+
+%%%------------------
+
+-spec add_variable_register(string(), tuple(), state()) -> state().
+add_variable_register(Var, Reg, ES = #es{debug_info = DbgInfo}) ->
+    VarsRegs = DbgInfo#debug_info.vars_registers,
+    Old = maps:get(Var, VarsRegs, []),
+    New = [Reg | Old],
+    NewDbgInfo = DbgInfo#debug_info{vars_registers = VarsRegs#{Var => New}},
+    ES#es{debug_info = NewDbgInfo}.
+
+-spec del_variable_register(string(), tuple(), state()) -> state().
+del_variable_register(Var, Reg, ES = #es{debug_info = DbgInfo}) ->
+    VarsRegs = DbgInfo#debug_info.vars_registers,
+    New = lists:delete(Reg, maps:get(Var, VarsRegs, [])),
+    NewDbgInfo = DbgInfo#debug_info{vars_registers = VarsRegs#{Var => New}},
+    ES#es{debug_info = NewDbgInfo}.
+
+-spec get_variable_register(string(), state()) -> tuple().
+get_variable_register(Var, #es{debug_info = #debug_info{vars_registers = VarsRegs}}) ->
+    case maps:get(Var, VarsRegs, [undefined]) of
+        []        -> undefined;
+        [Reg | _] -> Reg
+    end.
+
+%%%------------------
+
+debugger_resume(ES = #es{debug_info = Info = #debug_info{status = step}}) ->
+    ES#es{debug_info = Info#debug_info{status = break}};
+debugger_resume(ES = #es{debug_info = Info = #debug_info{status = {next, Stack}}}) ->
+    case length(ES#es.call_stack) == length(Stack) of
```

Comment:
I'm passing the entire call stack along with `{next, Stack}` but I'm not sure that this is needed. Possibly only the size of the call stack or the head of the call stack (if it's unique) should work.

Reply:
I got lost, why does it depend on the stack? Isn't it supposed to just execute until next `DBG_LOC`?